test/concatenation_tests.jl

```diff
@@ -0,0 +1,125 @@
+using MCMCChains
+using Test
+
+@testset "vcat" begin
+    chn = Chains(rand(10, 5, 2), ["a", "b", "c", "d", "e"], Dict(:internal => ["d", "e"]))
+    chn1 = Chains(rand(5, 5, 2), ["a", "b", "c", "d", "e"], Dict(:internal => ["a", "b"]))
+
+    # incorrect thinning
+    @test_throws ArgumentError vcat(chn, Chains(rand(2, 5, 2); thin = 2))
+
+    # incorrect names
+    @test_throws ArgumentError vcat(chn, Chains(rand(10, 5, 2), ["a", "b", "c", "d", "f"]))
+
+    # incorrect number of chains
+    @test_throws ArgumentError vcat(chn, Chains(rand(10, 5, 3), ["a", "b", "c", "d", "e"]))
+
+    # concate the same chain
+    chn2 = vcat(chn, chn)
+    @test chn2.value.data == vcat(chn.value.data, chn.value.data)
+    @test size(chn2) == (20, 5, 2)
+    @test names(chn2) == names(chn)
+    @test range(chn2) == 1:20
+    @test chn2.name_map == (internal = ["d", "e"], parameters = ["c", "b", "a"])
+    
+    chn2a = cat(chn, chn)
+    @test chn2a.value == chn2.value
+    @test chn2a.name_map == chn2.name_map
+    @test chn2a.info == chn2.info
+
+    chn2b = cat(chn, chn; dims = Val(1))
+    @test chn2b.value == chn2.value
+    @test chn2b.name_map == chn2.name_map
+    @test chn2b.info == chn2.info
+
+    chn2c = cat(chn, chn; dims = 1)
+    @test chn2c.value == chn2.value
+    @test chn2c.name_map == chn2.name_map
+    @test chn2c.info == chn2.info
+
+    # concatenate a different chain
+    chn3 = vcat(chn, chn1)
+    @test chn3.value.data == vcat(chn.value.data, chn1.value.data)
+    @test size(chn3) == (15, 5, 2)
+    @test names(chn3) == names(chn)
+    @test range(chn3) == 1:15
+    @test chn3.name_map == (internal = ["a", "b"], parameters = ["c", "e", "d"])
```

Comment:
Is this the intended behaviour? With the current implementation, in every concatenation only the `name_map` of the final argument is kept (even in case of `hcat`) if the chains use the same sections.

Reply:
I believe it was intended, but I don't think it's a particularly good idea. I suppose the ideal case would be to should construct the set addition version of the name map.

Reply:
But that only makes sense for `hcat`, doesn't it? There the different arguments are supposed to have unique names, so a set addition seems reasonable. However, for `vcat` and `chainscat` maybe we should just demand that all maps are equal?

Reply:
Sure, that's better.